### DISCUSSION TITLE
Grid API is dependent on properties set

### DISCUSSION THREAD

**Discussion by hanastasov** [IgniteUI/igniteui-angular#12099] [General]
At the moment, the grid API (probably other components as well) expects a different type of parameter, dependent on if a certain property is set to the grid.

Example: The `getRowByKey` method is implemented in a way that it accepts as a parameter both the primary key field value (if exists), or the row  data object itself, if primary key is not set:

![image](https://user-images.githubusercontent.com/14248932/190979217-2e1e18d6-d307-435d-afa9-0acc3ad709f3.png)

An implementation, that is agnostic to properties set to the grid, should always accepts the same type of parameter and not branch the logic inside based on the type of parameter sent.

**Comment by hanastasov** [comment]
If such a change should affect only public API, then it is only the `rowByKey` method that needs a slight refactoring.

**Reply by jberes** [reply]
I am not sure what rowByKey does, but for any event, whether a primary key is set or not, we should return the row data.

We should not have to use getRowData to retrieve cell values when doing direct grid interactions like selecting rows, selecting cells, etc.

getRowData s/b be for getting cell values in multi-selected rows?

**Reply by brianlagunas** [reply]
I don't see why I would ever have to call another grid method like `getRowData` inside a selection event. Even if I select a cell or multiple cells, I should have access to the row data of each selected cell.  I would only ever use a method like `getRowData` if I was working outside the grid events/interactions. Otherwise, if I make a selection, the event args should give me all the info about that selection.

Also, a selection event should **always** give me the row data. I should not be given some type of key and then forced to call another method off the grid to get the data of the selection. Not only is that highly inefficient, it's not intuitive.

**Reply by kdinev** [reply]
My 2 cents: As I've noted before, arguments in the events should be agnostic to grid configuration. The event shouldn't fire with different arguments depending on whether a grid-level prop like `primaryKey` is set or not. In the case of `rowSelectionChanging` and other such events, arguments like `newSelection`, `oldSelection` should be consistent. So, if we're passing reference to the selected data records without `primaryKey` set, then with `primaryKey` we should still have the same, not just a collection of the `primaryKey` values. 

Why? Because if I have the same grid configured, which over time receives a `primaryKey`, which it didn't have before, then I have to rewrite all of my already written logic to handle selection. 

The same applies to API methods. The `getRowData` method suggests it would retrieve underlying data for a provided row object. The method is not even necessary, since we already have a row object, which carries its data in its properties. Also `getRowData` expects a `primaryKey` to be provided. If I already have a reference to my data source and I know what my `primaryKey` field is, then why would I ever use the grid method? I can simply do this:

```typescript
this.data.find(m => m.primaryKey === 'uniqueValue`);
```

The `getRowByKey` suggests it would retrieve a row object when a key for that object is provided. This method actually assumes by the name that a `key` is provided to the grid, so it should work only in that scenario. This method makes sense, as the dev cannot construct a row interface themselves outside of the grid context.
